util/tracing/detect/recorder.go

```diff
@@ -26,37 +37,52 @@ type stubs struct {
 
 func NewTraceRecorder() *TraceRecorder {
 	tr := &TraceRecorder{
+		mu:        make(chan struct{}, 1),
+		done:      make(chan struct{}),
 		m:         map[trace.TraceID]*stubs{},
 		listeners: map[trace.TraceID]int{},
 	}
+	tr.mu <- struct{}{}
 
-	go func() {
-		t := time.NewTimer(60 * time.Second)
-		for {
-			<-t.C
-			tr.gc()
-			t.Reset(50 * time.Second)
-		}
-	}()
+	ctx, cancel := context.WithCancelCause(context.Background())
+	go tr.gcLoop(ctx)
+	tr.shutdown = cancel
 
 	return tr
 }
 
-func (r *TraceRecorder) Record(traceID trace.TraceID) func() []tracetest.SpanStub {
-	r.mu.Lock()
-	defer r.mu.Unlock()
+func (r *TraceRecorder) Record(ctx context.Context) func() []tracetest.SpanStub {
+	// Just a simplification for usage.
+	// If the trace recorder was never initialized, return no function.
+	// This allows the caller to know the trace recorder isn't set up
+	// more easily.
+	if r == nil {
+		return nil
+	}
+
+	spanCtx := trace.SpanContextFromContext(ctx)
+	if !spanCtx.IsValid() {
+		return emptyTraceSpans
```

Comment:
The original idea was because then this wouldn't trigger a "no trace recorder" error. I looked through the original path and it's probably just easier if we return nil everywhere than the stub empty function so I've removed this.